docs/external_objectives.rst

```diff
@@ -0,0 +1,318 @@
+        betapol = float(file.variables["betapol"][0])
+        betator = float(file.variables["betator"][0])
+        file.close()
+        return np.atleast_1d([betatot, betapol, betator])
+
+This function can then be used in an optimization with the ``ExternalObjective`` class.
+This objective must be initialized with the Equilibrium to be optimized, ``eq``, and the
+following other arguments:
+
+* ``fun``, the external function. In this example it is the function ``beta_from_file``.
+* ``dim_f``, the dimension of the output of the function ``fun``. In this example
+  ``dim_f=3`` because ``beta_from_file`` returns an array of size 3.
+* ``vectorized``, whether ``fun`` expects a single Equilibrium or a list of
+  Equilibria. Since the function ``beta_from_file`` takes a single Equilibrium as its
+  only positional argument, ``vectorized=False``.
+
+All other input parameters that the external function requires can be passed as keyword
+arguments to ``ExternalObjective``. In this example, ``path`` specifies the file name
+of the NetCDF file that the function writes/reads. Since ``vectorized=False``, the
+function ``beta_from_file`` will be evaluated sequentially when computing the finite
+differences (rather than in parallel), so the same file will be overwritten each time
+for the different Equilibria.
+
+::
+
+    from desc.objectives import ExternalObjective, ObjectiveFunction
+
+    objective = ObjectiveFunction(
+        ExternalObjective(
+            eq=eq,
+            fun=beta_from_file,
+            dim_f=3,
+            vectorized=False,
+            path="path/to/file.nc",
+        )
+    )
+
+Note that for this example, the same optimization objective could be accomplished
+naitively in DESC with automatic differentiation instead of finite differences as:
+
+::
+
+    from desc.objectives import GenericObjective, ObjectiveFunction
+
+    objective = ObjectiveFunction(
+        (
+            GenericObjective("<beta>_vol", thing=eq),
+            GenericObjective("<beta_pol>_vol", thing=eq),
+            GenericObjective("<beta_tor>_vol", thing=eq),
+        )
+    )
+
+
+TERPSICHORE objective
+---------------------
+
+This section walks through the implementation of wrapping the ideal MHD linear stability
+code TERPSICHORE, which is written in FORTRAN. It summarizes how the external function
+was written to call TERPSICHORE from DESC, and can be used as a template for wrapping
+other codes. The code shown here is abbreviated and slightly modified for brevity, but
+the full source code can be found in ``desc/experimental/_terpsichore.py``.
+
+The external function in this case is named ``terpsichore``, and takes the following
+arguments:
+
+* ``eq``, a list of Equilibria.
+* ``processes``, the maximum number of processes to use for multiprocessing.
+* ``path``, the path to the directory where the optimization script is being executed.
+* ``exec``, the file name of the TERPSICHORE executable.
```

Comment:
this should be the file path to the executable right? not just filename

Reply:
It is only the filename. The real documentation for the `TERPSICHORE` objective explains that `exec` "must be located in the directory specified by `path`". But I am being succinct here since it is just a tutorial template.